libnmstate/nm/context.py

```diff
@@ -62,12 +100,98 @@ def _del_client(self):
+        """
+        queue = self._fast_queue if fast else self._slow_queue
+        max_queue = FAST_ASYNC_QUEUE_SIZE if fast else SLOW_ASYNC_QUEUE_SIZE
+        if len(queue) >= max_queue:
+            logging.debug(
+                f"Async queue({max_queue}) full, waiting all existing actions "
+                "to be finished before registering more async action"
+            )
+            # TODO: No need to wait all finish, a single finished async is
+            #       sufficient.
+            self.wait_all_finish()
+
+        logging.debug(f"Async action: {action} started")
+        queue.add(action)
+
+    def finish_async(self, action, suppress_log=False):
+        """
+        Mark action(string) as finished.
+        """
+        self._last_async_finish_time = datetime.datetime.now()
+        if not suppress_log:
+            logging.debug(f"Async action: {action} finished")
+        self._fast_queue.discard(action)
+        self._slow_queue.discard(action)
+
+    def _action_all_finished(self):
+        return not (len(self._fast_queue) or len(self._slow_queue))
+
+    def _idle_timeout_cb(self, _user_data):
+        if self._error or self._action_all_finished():
+            return GLib.SOURCE_REMOVE
+        idle_time = datetime.datetime.now() - self._last_async_finish_time
+        if idle_time > datetime.timedelta(IDLE_TIMEOUT):
+            remaining_actions = self._slow_queue | self._fast_queue
+            self.fail(
+                NmstateTimeoutError(f"Action {remaining_actions} timeout")
+            )
+            return GLib.SOURCE_REMOVE
+        else:
+            return GLib.SOURCE_CONTINUE
+
+    def is_cancelled(self):
+        return self._cancellable.is_cancelled()
+
+    def fail(self, exception):
+        if not self._cancellable.is_cancelled():
+            if self._error:
+                logging.error(
+                    f"BUG: There is already a exception assigned: "
+                    f"existing: {self._error}, new exception {exception}"
+                )
+            self.cancellable.cancel()
+            self._del_timeout()
+            self._error = exception
+
+    def wait_all_finish(self):
+        """
+        Block till all async actions been marked as finished via
+        `finish_async()` or anyone failed by `fail()`.
+        """
+        self._last_async_finish_time = datetime.datetime.now()
+        if not self._action_all_finished():
+            self._timeout_source = GLib.timeout_source_new(
+                IDLE_CHECK_INTERNAL * 100
+            )
+            user_data = None
+            self._timeout_source.set_callback(self._idle_timeout_cb, user_data)
+            self._timeout_source.attach(self._context)
+
+            while not self._action_all_finished() and not self._error:
+                self.context.iteration(True)
+            self._del_timeout()
+
+        if self._error:
+            raise self._error
+
+    def get_nm_dev(self, iface_name):
```

Comment:
Why the `context` knows about devices?

Reply:
The device is query from the cache of `NM.Client`.
And I dislike the `nm.device.get_device()` require user to provide the `NM.Client` instance every
time. (the same as nm/connection.py and nm/action_conneciton.py, but this one is widely used.)

Reply:
It is not less ugly to let context know about all NM functions (like this one).
As I do not have a better solution to suggest, ugly it will remain.